### PR TITLE
(#Closes 2196) Intrinsic implementation for fparser2, container symbol and fortran backend

### PR DESCRIPTION
**Pull request by LonelyCat124**
This should be implemented - I maybe missing a specific test - let me know if so and I can add it.

### CODE REVIEW

**Comment by LonelyCat124** [comment]
Ready for a first review.

File: src/psyclone/psyir/symbols/containersymbol.py [hunk 1]
```diff
@@ -163,6 +169,27 @@ def wildcard_import(self, value):
...
+        '''
+        :returns: whether or not this module is an intrinsic module.
+        :rtype: bool
+        '''
+        return self._intrinsic
+
+    @intrinsic.setter
+    def intrinsic(self, value):
+        '''
+        Set whether or not this is an intrinsic module import.
```
Comment (sergisiso) [inline src/psyclone/psyir/symbols/containersymbol.py:183]: You can get rid of the property description of its the same as the first param.

File: src/psyclone/psyir/symbols/containersymbol.py [hunk 2]
```diff
@@ -163,6 +169,27 @@ def wildcard_import(self, value):
...
+        :param bool value: whether or not this is an intrinsic module.
+
+        :raises TypeError: if the supplied `value` is not a bool.
+        '''
+        if not isinstance(value, bool):
+            raise TypeError(f"intrinsic must be a bool but got: "
+                            f"'{type(value).__name__}'")
```
Comment (sergisiso) [inline src/psyclone/psyir/symbols/containersymbol.py:190]: Start with capital letter and end with punctuation.

File: src/psyclone/psyir/symbols/containersymbol.py [hunk 3]
```diff
@@ -63,6 +63,7 @@ def __init__(self, name, wildcard_import=False, **kwargs):
 
         self._reference = None
         self._has_wildcard_import = False
+        self._intrinsic = False
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/symbols/containersymbol.py:66]: And this is a bit subjective but for this kind of bools I prefer if they start with "is", e.g. `_is_intrinsic`, the same for the arguments and properties. IMO it leads to a more readable code.

File: src/psyclone/tests/psyir/frontend/fparser2_test.py [hunk 4]
```diff
@@ -1926,6 +1926,24 @@ def test_process_use_stmts_with_accessibility_statements(parser):
...
 
 
+def test_intrinsic_use_stmt(parser):
+    ''' Tests that intrinsic value is set correctly for an intrinsic module
+    use statement.'''
+    processor = Fparser2Reader()
+    reader = FortranStringReader('''
+        module test
+            use, intrinsic :: ieee_arithmetic, only: isnan =>ieee_is_nan
+            use mymod
```
Comment (sergisiso) [inline src/psyclone/tests/psyir/frontend/fparser2_test.py:1936]: For completeness you can also test the `use, non_intrinsic :: ...` import  which is the only other possible value and it should be the same as default here (no need to reproduce it in the backend, as it is equivalent to no attribute)

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 5]
```diff
@@ -581,6 +581,10 @@ def test_fw_gen_use(fortran_writer):
     assert "use my_module, only : dummy1=>orig_name, my_sub" not in result
     assert "use my_module\n" in result
 
+    container_symbol.intrinsic = True
+    result = fortran_writer.gen_use(container_symbol, symbol_table)
+    assert "intrinsic" in result
```
Comment (sergisiso) [inline src/psyclone/tests/psyir/backend/fortran_test.py:586]: Can you compare the test of the whole use statement please?

File: src/psyclone/psyir/backend/fortran.py [hunk 6]
```diff
@@ -483,15 +483,20 @@ def gen_use(self, symbol, symbol_table):
                 # This variable is not renamed.
                 only_list.append(dsym.name)
 
+        intrinsic_str = ""
+        if symbol.intrinsic:
+            intrinsic_str = "intrinsic "
```
Comment (sergisiso) [inline src/psyclone/psyir/backend/fortran.py:488]: Is this missing the comma-space before intrinsic?
image

**Review by sergisiso — CHANGES_REQUESTED** [review]
Thanks for tackling this @LonelyCat124, it not good when we just ignore some fparser attribute. There are only minor comments to clean up the code.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 7]
```diff
@@ -1691,6 +1691,10 @@ def _process_use_stmts(parent, nodes, visibility_map=None):
                     f"Expected the parse tree for a USE statement to contain "
                     f"5 items but found {len(decl.items)} for '{text}'")
 
+            intrinsic = False
+            if decl.items[0] is not None and str(decl.items[0]) == "INTRINSIC":
+                intrinsic = True
+
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/frontend/fparser2.py:1697]: Added

File: src/psyclone/psyir/symbols/containersymbol.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/symbols/containersymbol.py:66]: Fixed

File: src/psyclone/psyir/symbols/containersymbol.py (hunk 1 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/symbols/containersymbol.py:183]: Done

File: src/psyclone/psyir/symbols/containersymbol.py (hunk 2 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/symbols/containersymbol.py:190]: Added

File: src/psyclone/psyir/symbols/containersymbol.py (hunk 1 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/symbols/containersymbol.py:183]: Actually i'm not sure I understand what you mean here? Is this docs related?

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 5 above)
Comment (LonelyCat124) [reply src/psyclone/tests/psyir/backend/fortran_test.py:586]: Added

File: src/psyclone/psyir/backend/fortran.py (hunk 6 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/fortran.py:488]: Yes - I've fixed this now

File: src/psyclone/tests/psyir/frontend/fparser2_test.py (hunk 4 above)
Comment (LonelyCat124) [reply src/psyclone/tests/psyir/frontend/fparser2_test.py:1936]: Added

**Comment by LonelyCat124** [comment]
@sergisiso should be ready for another look now I think - one comment I didn't address as I'm unclear what you meant - I tried to remove `@property` but then the code didn't work.

**Review by sergisiso — APPROVED** [review]
@LonelyCat124 All the suggestions have been addressed, everything is tested, and the coverage is complete. If the integration test also passes this is ready to merge.

**Pull request merged**